tensorflow_probability/python/distributions/johnson_su.py

```diff
@@ -0,0 +1,287 @@
+               gamma,
+               delta,
+               loc,
+               scale,
+               validate_args=False,
+               allow_nan_stats=True,
+               name=None):
+    """Construct Johnson's SU distributions.
+
+    The distributions have shape parameteres `delta` and `gamma`, mean `loc`,
+    and scale `scale`.
+
+    The parameters `delta`, `gamma`, `loc`, and `scale` must be shaped in a way
+    that supports broadcasting (e.g. `gamma + delta + loc + scale` is a valid
+    operation).
+
+    Args:
+      gamma: Floating-point `Tensor`. The shape parameter(s) of the
+        distribution(s).
+      delta: Floating-point `Tensor`. The shape parameter(s) of the
+        distribution(s). `delta` must contain only positive values.
+      loc: Floating-point `Tensor`. The mean(s) of the distribution(s).
+      scale: Floating-point `Tensor`. The scaling factor(s) for the
+        distribution(s). Note that `scale` is not technically the standard
+        deviation of this distribution but has semantics more similar to
+        standard deviation than variance.
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`,
+        statistics (e.g., mean, mode, variance) use the value '`NaN`' to
+        indicate the result is undefined. When `False`, an exception is raised
+        if one or more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+
+    Raises:
+      TypeError: if any of gamma, delta, loc and scale are different dtypes.
+    """
+    parameters = dict(locals())
+    with tf.name_scope(name or 'JohnsonSU') as name:
+      dtype = dtype_util.common_dtype([gamma, delta, loc, scale], tf.float32)
+      self._gamma = tensor_util.convert_nonref_to_tensor(
+          gamma, name='gamma', dtype=dtype)
+      self._delta = tensor_util.convert_nonref_to_tensor(
+          delta, name='delta', dtype=dtype)
+      self._loc = tensor_util.convert_nonref_to_tensor(
+          loc, name='loc', dtype=dtype)
+      self._scale = tensor_util.convert_nonref_to_tensor(
+          scale, name='scale', dtype=dtype)
+      dtype_util.assert_same_float_dtype((self._gamma, self._delta,
+                                          self._loc, self._scale))
+
+      norm_shift = invert_bijector.Invert(
+          shift_bijector.Shift(shift=self._gamma, validate_args=validate_args)
+      )
+
+      norm_scale = invert_bijector.Invert(
+          scale_bijector.Scale(scale=self._delta, validate_args=validate_args)
+      )
+
+      sinh = inline_bijector.Inline(
+          forward_fn=tf.sinh,
+          inverse_fn=tf.asinh,
+          forward_log_det_jacobian_fn=lambda x: tf.math.log(tf.math.cosh(x)),
```

Comment:
Thanks for the review! Could you clarify a bit more on this? How `abs` appears here? I've put `x + tf.math.softplus(-2. * x) - np.log(2.)` into the code and tests are passing.

Reply:
I think you should be able to use `tfp_math.log_cosh` now which also handles x close to zero:
https://cs.corp.google.com/piper///depot/google3/third_party/py/tensorflow_probability/python/bijectors/sinh_arcsinh.py?rcl=300816317&l=143

The reason I was mentioning `abs` before, was so as to avoid numerical instability (if x is negative, then softplus(-2 * x) is something like -2 * x + softplus(2 * x). Thus we would get something like x - 2 * x. We can avoid such a cancellation if we use abs).

Reply:
Replaced with `tfp_math.log_cosh` and now have `JohnsonSUTest.testFiniteGradientAtDifficultPoints` test failing with `nan` last 2 gradients. Quick research brought no results. Will try to figure out what's going on.

Reply:
tfp.math.log_cosh is failing due to https://github.com/tensorflow/tensorflow/issues/38349

`tf.math.log1p(-tf.square(x) / 6.)` seems to be evaluated for big `x` eventhough this `where` branch is inactive.

Do you think it would be better to wait for the issue addressed in TF and then migrate to `log_cosh` here or merge current implementation and migrate once the issue is fixed? Any other options?

Reply:
@srvasude TF guys said the issue won't be fixed. Are you fine if I apply following patch?
```
--- a/tensorflow_probability/python/math/generic.py
+++ b/tensorflow_probability/python/math/generic.py
@@ -505,10 +505,12 @@ def log_cosh(x, name=None):
     x = tf.math.abs(x)
     logcosh = x + tf.math.softplus(-2 * x) - np.log(2).astype(numpy_dtype)
     bound = 45. * np.power(np.finfo(numpy_dtype).tiny, 1 / 6.)
+    # Workaround for https://github.com/tensorflow/tensorflow/issues/38349
+    safe_x = tf.where(x <= bound, x, 1.)
     return tf.where(
         x <= bound,
         tf.math.exp(
-            tf.math.log(x) + tf.math.log1p(-tf.square(x) / 6.)),
+            tf.math.log(safe_x) + tf.math.log1p(-tf.square(safe_x) / 6.)),
         logcosh)
```

Reply:
Just submitted a fix to the gradient, please try now (I just overrode the gradient to be tf.tanh(x)).

Reply:
It works now, thanks!